pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -0,0 +1,272 @@
+ * Hence, we rewrite such expressions into an equivalent expression whose LHS and RHS are of the same datatype.
+ *
+ * Simple predicate examples:
+ *  1) WHERE "intColumn = 5.0"  gets rewritten to "WHERE intColumn = 5"
+ *  2) WHERE "intColumn != 5.0" gets rewritten to "WHERE intColumn != 5"
+ *  3) WHERE "intColumn = 5.5"  gets rewritten to "WHERE false" because INT values can not match 5.5.
+ *  4) WHERE "intColumn = 3000000000 gets rewritten to "WHERE false" because INT values can not match 3000000000.
+ *  5) WHERE "intColumn != 3000000000 gets rewritten to "WHERE true" becuase INT values always not equal to 3000000000.
+ *
+ * Compound predicate examples:
+ *  6) WHERE "intColumn1 = 5.5 AND intColumn2 = intColumn3"
+ *       rewrite to "WHERE false AND intColumn2 = intColumn3"
+ *       rewrite to "WHERE intColumn2 = intColumn3"
+ *  7) WHERE "intColumn1 != 5.5 OR intColumn2 = 5000000000" (5000000000 is out of bounds for integer column)
+ *       rewrite to "WHERE true OR false"
+ *       rewrite to "WHERE true"
+ *       rewrite to query without any WHERE clause.
+ *
+ * When entire predicate gets rewritten to false (Example 3 above), the query will not return any data. Hence, it is
+ * better for the Broker itself to return an empty response rather than sending the query to servers for further
+ * evaluation.
+ */
+public class NumericalFilterOptimizer implements FilterOptimizer {
+
+  private static final Expression TRUE = RequestUtils.getLiteralExpression(true);
+  private static final Expression FALSE = RequestUtils.getLiteralExpression(false);
+
+  @Override
+  public FilterQueryTree optimize(FilterQueryTree filterQueryTree, @Nullable Schema schema) {
+    // Don't do anything here since this is for PQL queries which we no longer support.
+    return filterQueryTree;
+  }
+
+  @Override
+  public Expression optimize(Expression expression, @Nullable Schema schema) {
+    ExpressionType type = expression.getType();
+    if (type != ExpressionType.FUNCTION) {
+      // Not a function, so we have nothing to rewrite.
+      return expression;
+    }
+
+    Function function = expression.getFunctionCall();
+    List<Expression> operands = function.getOperands();
+    String operator = function.getOperator();
+    if (operator.equals(FilterKind.AND.name()) || operator.equals(FilterKind.OR.name())) {
+      // One of the operands may be an EQUALS or NOT_EQUALS function so recursively traverse the expression tree to see
+      // if we find an EQUALS or NOT_EQUALS function to rewrite.
+      operands.forEach(operand -> optimize(operand, schema));
+
+      // We have rewritten the child operands, so rewrite the parent if needed.
+      return optimizeCurrent(expression);
+    } else if (operator.equals(FilterKind.EQUALS.name()) || operator.equals(FilterKind.NOT_EQUALS.name())) {
+      // Verify that LHS is a numeric column and RHS is a numeric literal before rewriting.
+      Expression lhs = operands.get(0), rhs = operands.get(1);
```

Comment:
(nit) Don't put two declaration on the same line

Reply:
Done